kebechet/services.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+# Kebechet
+# Copyright(C) 2018, 2019 Kevin Postlethwait
+#
+# This program is free software: you can redistribute it and / or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Provides abstraction of service utility functions."""
+
+
+import requests
+import tempfile
+import logging
+
+from .enums import ServiceType
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _keep_slug(slug: str):
+    return slug
+
+
+def _encode_fslash(slug: str):
+    return slug.replace("/", "%2F")
+
+
+# NOTE: config_url could be generated using construct_raw_file_url but I feel this makes the code easier to follow
+services = {
+    "github": {
+        "service_type": ServiceType.GITHUB,
+        "config_url": "https://raw.githubusercontent.com/{slug}/{branch}/.kebechet.yaml",
+        "auth": {"header": "Authorization", "value": "token {token}"},
+        "slug_method": _keep_slug,
+    },
+    "gitlab": {
+        "service_type": ServiceType.GITLAB,
+        "config_url": "https://gitlab.com/api/v4/projects/{slug}/repository/files/.kebechet.yaml/raw?ref={branch}",
+        "auth": {"header": "Private-Token", "value": "{token}"},
+        "slug_method": _encode_fslash,
+    },
+}
+
+
+class Service():
+    """Allows for abstraction of multiple different service with easy support of my by changing `services`."""
+
+    def __init__(self, service, url, token, branch="master"):
+        """Initialize a Service object, calls constructor functions `slug_method` and `_generate_slug`."""
+        if service not in services.keys():
+            _LOGGER.warning("Service not supported")
+            raise ValueError(f"{service} is not supported at this time")
+        self.service = services[service]["service_type"]
+        _LOGGER.info(f"{service} service detected")
```

Comment:
```python
_LOGGER.info("%s service detected", service)
```

... so strings are expanded lazily. Here and in other parts. Some linters might complain.

Reply:
I was following conventions that I found elsewhere in the code like here:
https://github.com/thoth-station/kebechet/blob/dca3392af69fb9e7726325716966cbc7f221d90e/kebechet/config.py#L170-L172

Should all code be updated to this % style?

Reply:
It would be better, but don't bother with old code. Let's make sure the newly added one is better :)

Reply:
Okay